### DISCUSSION TITLE
What parameters does a png need to meet to be converted to pdf 1.3?

### DISCUSSION THREAD

**Discussion by developertrinidad08** [ImageMagick/ImageMagick#7680] [Help]
### ImageMagick version

7.1.1-38

### Operating system, version and so on

Manjaro 

### Description

Hello everyone...!!!
I'll put you in context. I'm making an app that sends a PDF to print from a client and most printers only support PDF v1.3 and as in IM if the image has certain parameters it normally converts it to 1.3. **I want to know what parameters it needs to meet?**
At the moment, by limiting the color space to CMYK and removing the alpha channels I always get version 1.3 when converting it to PDF.
But before putting it into production I would like to isolate these parameters as much as possible to be 99.99% sure that I will always get the version I need. 

I currently use the following commands to get the version I need:

```
 cmd0 = "convert #{img} -alpha off -colorspace CMYK output#{random}.png "
 cmd1 = "convert output#{random}.png -units PixelsPerInch -alpha off -density #{dpi} PDF:output#{random}.pdf
```"

Basically I want to know what other commands to add to cmd0 to make sure magick always returns a pdf v1.3

**Comment by fmw42** [comment]
You do not need two commands. You can do all that in one command. 

`magick #{img} -alpha off -colorspace CMYK -units PixelsPerInch -alpha off -density #{dpi} PDF:output#{random}.pdf`

Note that on IM 7, you should use magick in place of convert.

**Reply by developertrinidad08** [reply]
Thank you very much for taking the time to respond
Regarding your recommendations
These are things that I take into account...
But I don't have it with just one line because previously I didn't pay attention to the -alpha off when I converted a png to pdf (I really didn't try it again, most likely I wrote something wrong) and it just stayed like that.
And I keep the convert command because I have a server that uses one of the first versions of IM 6 and it is more convenient for me to continue using that command than to update each server where I will put my app in the future.

**Reply by developertrinidad08** [reply]
But something I would really like to know is:
apart from the fact that it doesn't have to have transparency and the color space must be CMYK (at least this worked for me in my tests)
what else do I need to "limit" to make sure that my PDF version is always 1.3? So far with these 2 limitations I'm not really having any problems, but I wouldn't be surprised if they call me to tell me that the printers don't print (happens with v1.7) or print blank pages (happens with v1.4).

**Reply by fmw42** [reply]
Then why did you write IM7 in your first post. You should have reported on the IM 6 forum.

**Reply by developertrinidad08** [reply]
It is only one particular server that uses this version, the others use IM 7. Simply using convert ensures that if I install my app and the IM is an older version, it also works and I don't have to have a version for IM6 and another for IM7.
or I don't have to update the server until I get to version 7 of IM. 

At the moment this works correctly for me whether I install in an IM6 or IM7 version, although the IM6 version always returns a V1.3 PDF without me adding anything to the command. In IM7, this is where I have the problems of returning other versions. 

which is really only a problem for my particular case where I hardly have a printer that supports any other version of PDF other than 1.3

**Reply by fmw42** [reply]
IM 6 is deprecated. So you should now with current versions get a warning.

**Reply by developertrinidad08** [reply]
If I get the warning... but since it works, at least at this stage of development I kept using it without worrying too much. But I think the best practice, based on what you're telling me, would be to put an if that checks the version and uses magick in IM7 and converts to IM6. My ideal would really be to always use IM7, but sometimes I have to work with very old servers that have applications that depend on old libraries, and updating something carelessly could cause things that work correctly to stop working.

**Comment by snibgo** [comment]
From coders/pdf.c (which I have already told you about): When writing a PDF, the default PDF version is 1.3.

If any image has alpha, the version is 1.4.

If compression is JPEG2000, the version is 1.5.

If the filetype (file extension) is PDFA, the version is 1.6.

If any image has an ICC color profile, the version is 1.7.

If your printer cant handle ICC colour profiles, I suggest you get a new printer. Converting to CMYK without a ICC colour profile is a really bad idea.

IM `convert` is deprecated, so using it is a really bad idea. Instead, use `magick`.
